src/Disposable.js

```diff
@@ -0,0 +1,98 @@
+module.exports = function() {
+    
+    function Disposable(dispose) {
+        !!dispose && (this._dispose = dispose);
+        this.disposed = false;
+        this.length = 0;
+    }
+    
+    Disposable.prototype.dispose = function() {
+        if(!this.disposed) {
+            var result = true, f, xs, i, n;
+            (f = this._dispose) && (result = f.call(this));
+            if(!!result || result === void 0) {
+                this.disposed = true;
+                delete this._dispose;
+                while(xs = this._disposables) {
+                    i = -1;
+                    n = xs.length;
+                    delete this._disposables;
+                    while(++i < n) {
+                        xs[i].dispose();
+                    }
+                }
+                this.length = 0;
+            }
+        }
+        return result;
+    };
+    
+    Disposable.prototype.add = function () {
+        
+        var argsIdx = -1;
+        var argsLen = arguments.length;
+        var disposables = new Array(argsLen);
+        while(++argsIdx < argsLen) { disposables[argsIdx] = arguments[argsIdx]; }
+        
+        var xs = this._disposables || (this._disposables = []),
```

Comment:
This will also cause deoptimization in V8. Adding members in an unknown order causes creation of a new "hidden class".

We should have a Disposable and a CompositeDisposable which inherits from it.